### ISSUE TITLE
New install and import APIs

### ISSUE THREAD

**Issue by hoodmane** [pyodide/pyodide#1940]
## 🚀 Feature

We should add an API like `pyodide.import("package_name");` that imports a package and returns it as a `PyProxy`. Possibly including a variant like `pyodide.import("package_name", ["name1", "name2"]);` or `pyodide.import("package_name", "*");`.

We should also add an API like `pyodide.installFiles` that wraps the new API in #1935.

### Motivation

We should work towards encouraging people to define Python functions and then call Python functions from Javascript, rather than using templating. `pyodide.runPython` is like `eval`, in general it's considered an antipattern to eval a templated string. But I think our current status makes it unclear to new users how to avoid it.

We have existing PRs to add Python home directory #1936 and add good APIs to directly install files into the Python home directory #1935. 

### Pitch

We should move towards thinking that `pyodide.runPython` is not the default way to run Python code. Instead, we should encourage that the users to write python source files (with all the tooling benefits that entails, like code highlighters, formatters, type checkers, intellisense). Then they should use an API to install their code into the file system (future work to add this to the js api). Then they should use this import API to import their package, and then directly call the APIs from Javascript.

So they would:
1. Call `pyodide.install("url to archive or single source file");` which would install the file in their home directory.
2. Say `let my_package = pyodide.import("my_package");`
3. Use their package directly e.g., `my_package.some_api(37, "some_arg");`

This instead of `pyodide.runPython` should be the new best practice for using `pyodide`.

**Comment by ryanking13** [comment]
> We should work towards encouraging people to define Python functions and then call Python functions from Javascript, rather than using templating.

Yeah, definitely agree.

After APIs like `pyodide.install` are added, we could also encourage people to define Python packages that do the main job and use `pyodide.runPython` just for an entrypoint.

1. `pyodide.install("main_package")`
2. `pyodide.runPython("import main_package; minimal_bootstrap(); main_package.do_main_stuffs()")`

**Comment by hoodmane** [comment]
I'm concerned that there is currently a slippery slope from 
`pyodide.runPython("main_package.do_main_stuffs()")`
to
```pyodide.runPython(`main_package.do_main_stuffs({"a" : ${some_value}, "b" : ${some_other_value}})`)```
when it would be better would be:
```js
let main_package = pyodide.runPython("import main_package; main_package");
main_package.do_main_stuffs(pyodide.toPy({a : some_value, b : some_other_value}));
```
so in particular I would like to privilege `pyodide.runPython("import pkg; pkg");` with a dedicated API.

**Comment by ryanking13** [comment]
Yeah, I agree `pyodide.import` would be better than doing stuffs inside `pyodide.runPython`.

**Comment by rth** [comment]
Thanks for the proposal!

> We should add an API like pyodide.import("package_name"); that imports a package and returns it as a PyProxy

Doesn't this mean to some extent reviving [pyodide.pyimport](https://pyodide.org/en/stable/usage/api/js-api.html#pyodide.pyimport)  we deprecated in favor of `pyodide.globals.get('key') `? Though pymport only passed objects that were already imported. We can also revert that removal but change the behavior . The`pyimport` name is better IMO than importPythonModule.

 > We should also add an API like pyodide.installFiles that wraps the new API in #1935.

For users I think it might be simpler to add support in micropip to install zip files (normal pip does do that, though it also runs setup.py inside the zip) with prefix option. Otherwise there is a risk they would get lost in all our installation methods.

**Comment by hoodmane** [comment]
Yeah I like `pyimport` better than `importPythonModule`. This works completely different than the original method though:
```js
let statistics = pyodide.pyimport("statistics")
statistics.mean([1,2,3]); // 2.0
pyodide.globals.get("statistics"); // undefined
```
and
```js
pyodide.runPython("x=2");
pyodide.globals.get("x"); // 2
pyodide.pyimport("x"); // ModuleNotFoundError
```

**Comment by hoodmane** [comment]
> For users I think it might be simpler to add support in micropip to install zip files (normal pip does do that, though it also runs setup.py inside the zip) with prefix option. Otherwise there is a risk they would get lost in all our installation methods.

I was hoping to avoid making them load micropip, especially since we don't need any of this wheel processing `packaging` stuff we just want to unzip a zip file into the home directory.

**Comment by rth** [comment]
> This works completely different than the original method though:

Indeed. IMO `pyimport` name makes much more sense for the proposed usage than the one we had before.
We deprecated it in earlier version and it's going to be removed in 0.19, so it's a breaking change for people who still use despite of the deprecation. If we re-introduce it now with this new functionality, I wouldn't see it as so much of a problem (it's an API change in any case). Unless you find it too confusing for users or we find another short name for this import functionality.

> I was hoping to avoid making them load micropip, especially since we don't need any of this wheel processing packaging stuff we just want to unzip a zip file into the home directory.

I see you point. I'm not opposed to having for semi-private api in pyodide for this. The question is what should be the recommended way of installing stuff in Pyodide. We would have,
 - pyodide.loadPackage (Js library). To install Pyodide packages.
 - micropip.install (separate Python library). To install wheels
 - pyodide.installFiles (Python library), or some similar name. To install zip files.

It's very confusing, even for someone who has followed these discussions. I find it would be better to say you install everything with micropip (most users need external packages anyway). Then if you want to do specific optimizations you can use lower level and semi-private API of loadPackage or installFiles. I'm just not sure how to make this more obvious in our documentation. Maybe `pyodide._installFiles` but still have it in the documentation? Nor sure.

> Possibly including a variant like pyodide.import("package_name", ["name1", "name2"]); or pyodide.import("package_name", "*");

How would those work?  if you import name1 and name2, does it mean that the return type is a tuple or does it internally define name1 and name2 js variables? Less sure about this, importing the module and storing it some variable is not so bad from the namespace perspective. For instance, having `pyodide.import("package_name", "*")`  define JS variables for each Python object that would be imported with `from package_name import *` feels like a pretty bad practice to me.

**Comment by rth** [comment]
Also do you propose to add the install API to the JS or Python package? Users from Python would also need to be able to easily install zip archives. But if it's there, running `pyodide.runPython("pyodide.installFiles('..')")` from JS is not ideal.

**Comment by ryanking13** [comment]
> * pyodide.loadPackage (Js library). To install Pyodide packages.
> * micropip.install (separate Python library). To install wheels
> * pyodide.installFiles (Python library), or some similar name. To install zip files.
> 
> It's very confusing, even for someone who has followed these discussions. I find it would be better to say you install everything with micropip (most users need external packages anyway).

For Python package installation, my opinion is that `micropip` should be a one and only name for both from JS and from Python.
Because people from Python domain will be familiar with `pip install` to install packages.

But for relative import (which is not actually installing packages), it's a bit complicated I guess. Because we don't do `pip install` to import a file in native environments. Doing `micropip.install` to import a "file" is a little confusing to me.

**Comment by rth** [comment]
> But for relative import (which is not actually installing packages), it's a bit complicated I guess. Because we don't do pip install to import a file in native environments. Doing micropip.install to import a "file" is a little confusing to me.

Installing separate files might indeed need a different API, but as far as I understand the proposal here is to install archives so far? Or would this also be used for relatives imports? Wheels are pretty much zip files + some dependency resolution, with a different installation folder.

**Comment by ryanking13** [comment]
> Or would this also be used for relatives imports?

I thought install_archive's usage is to prevent a bunch of requests to download individual python files, and it's mainly for relative imports. (https://github.com/pyodide/pyodide/discussions/1876#discussioncomment-1560212)

Maybe I was misunderstanding the usage?

**Comment by hoodmane** [comment]
> IMO `pyimport` name makes much more sense for the proposed usage than the one we had before.

I'm +1 for using `pyimport` if you like it. I think since it was deprecated it's fine.

> I find it would be better to say you install everything with `micropip` (most users need external packages anyway).

I agree that using `micropip` for everything would be simpler. I don't like it if people have to `await pyodide.loadPackage('micropip');` first though. I think if we are going to recommend micropip for everything, we should include it when `fullStdLib` is true.

**Comment by hoodmane** [comment]
> >Possibly including a variant like pyodide.import("package_name", ["name1", "name2"]); or pyodide.import("package_name", "*");
>
> How would those work?

I read how `importlib.__import__` works and I don't think this is a good idea anymore.

**Comment by grimmer0125** [comment]
@ryanking13, @hoodmane  I saw this issue discussion and #1917 and #1935. I think there are 2 similar but a little different uses cases. 
1. import (multiple) Python file(s) at once, like **importing a Python package**. Could be a single file or multiple files.
2. import (multiple) Python files(s) at once and run it, like **running a Python main program**. Could be a single file or multiple files.

If the 2nd case is also your target in the long term, I would suggest having something like `if __name__ == '__main__'` to distinguish the main entry file and other imported files, there might be some usage that people want to avoid some code been executed. Another benefit is that it will not rely on the expression of the final line as the returned value when using these new Pyodide API. Specifying the main file name in the API is the equivalent thing, and either one is OK or co-exist. 

Personally, I think it will be great as long as any of these two is supported. 

BTW, I'm not sure I'm pasting here is a good suitable place or it is better to continue https://github.com/pyodide/pyodide/discussions/1876, or other places.

**Comment by ryanking13** [comment]
> BTW, I'm not sure I'm pasting here is a good suitable place or it is better to continue #1876, or other places.

We may better separate discussion of two APIs:

- Install API: `pyodide.installFiles` + other ideas
  - How to download external files, and make them importable.
- Import API: `pyodide.pyimport`
  - Possible substitute of `pyodide.runPython`

For Install API, It would be better to use #1917 or #1876 instead. I will move some previous comments in this issue if needed.
For Import API, I guess this issue is a good place to discuss?

**Comment by alexmojaki** [comment]
Having an `installFiles` function would be great. For reference, here's what happens in futurecoder:

1. A tarfile containing all the main Python code is fetched from a URL in JS.
2. A [small bootstrap python file](https://github.com/alexmojaki/futurecoder/blob/master/frontend/src/load.py) is bundled directly into the JS source as a string.
3. The Python source string is run using `runPython`.
4. The function defined in that string is fetched with `pyodide.globals.get`.
5. The Python function is called with a Buffer from step 1.
6. It extracts the tarfile into the virtual filesystem, adds to `sys.path`, and imports the entrypoint function of the extracted code.

All the JS code is [here](https://github.com/alexmojaki/futurecoder/blob/master/frontend/src/Worker.js).

I'd love to replace this with something official from Pyodide and to know that others don't have to figure this stuff out themselves. I've seen articles by people who thought that Pyodide only allows running one file, not complex projects with imports. It seems like what's being proposed here would handle all of this in one or two function calls.

Note:

1. I use a tarfile instead of a zip so that decompression can be handled by the browser instead of Python's zipfile module.
2. I don't use micropip in this process, just JS `fetch` and the Python tarfile module.
3. Downloading the tarfile happens in parallel to loading the bootstrap code, i.e. loading a few Python standard library modules. If downloading only started after loading the relevant Python code that would be a pity, but probably not a dealbreaker. https://github.com/ryanking13/pyodide-importer looks like too much though, requiring installing and loading micropip, then installing and loading `pyodide_importer`, and only starting the archive download after that.
4. The tarfile includes my code as well as external packages, which is why micropip isn't needed. micropip probably wouldn't be able to install all those dependencies anyway. The tarfile is generated by [this script](https://github.com/alexmojaki/futurecoder/blob/master/core/generate_static_files.py) which archives my own code plus dependencies in a local virtualenv. This is something else users could probably use help with. Just a thought, probably out of scope here.

**Comment by hoodmane** [comment]
Thanks for the input @alexmojaki. I will prototype a `pyodide.installFiles` soon and see if I can get enthusiasm from the other Pyodide developers. 

> If downloading only started after loading the relevant Python code that would be a pity, but probably not a dealbreaker.

Perhaps you could fetch the things yourself? What if we made `pyodide.installFiles` take either:
(1) a URL or (2) a promise that resolves to a Fetch `Response` object.
Then if you want to prefetch stuff, you just initiate the fetch yourself and hand in the promise when Pyodide is ready for it?

**Comment by hoodmane** [comment]
For example, what about:
```
async function main(){
   let tarPromise = fetch("myarchive.tar.gz");
   let pyodide = await loadPyodide({ /* options here */});
   await pyodide.installFiles(tarPromise, installDirectory);
   let my_pkg = pyodide.pyimport("my_package");
   my_pkg.main();
}
```

**Comment by ryanking13** [comment]
> Perhaps you could fetch the things yourself? What if we made `pyodide.installFiles` take either:
> (1) a URL or (2) a promise that resolves to a Fetch `Response` object.
> Then if you want to prefetch stuff, you just initiate the fetch yourself and hand in the promise when Pyodide is ready for it?

Or maybe a callback style?

```js
let tarPromise = fetch("myarchive.tar.gz");
let pyodide = await loadPyodide({ /* options here */});
tarPromise.then((response) => pyodide.installFiles(
   await pyodide.installFiles(response, installDirectory);
   ...
)
```

**Comment by hoodmane** [comment]
I figure that since `installFiles` has to be async anyways (because `response.arrayBuffer` is itself async) there is no reason we can't just accept a promise that resolves to a fetch response as the first argument. Just have to stick an extra `await` in.

I suppose that we could also make a synchronous API `pyodide.installFilesFromArrayBuffer(arrayBuffer, format, [directory])`, not sure anyone would need it though.

**Comment by ryanking13** [comment]
I agree that `installFiles` has to be async.
I think the possible advantage of passing ArrayBuffer instead of Promise is that exceptions raised by `fetch` can be easily handled outside of `pyodide.installFiles`. But I'm not a very JS person, so I'm not sure it's worth it, maybe @grimmer0125 can share his idea?

**Comment by alexmojaki** [comment]
> I think the possible advantage of passing ArrayBuffer instead of Promise is that exceptions raised by `fetch` can be easily handled outside of `pyodide.installFiles`. But I'm not a very JS person, so not sure it's worth it, maybe @grimmer0125 can share his idea?

I agree, allowing an ArrayBuffer (or a Promise<ArrayBuffer>) is more flexible and customisable. Maybe the user doesn't want to use fetch at all, or they want to check the status code, or do retries, etc. I don't think I *need* these things, but I wouldn't personally mind having to add the extra `response.arrayBuffer()` line or whatever. And you could probably just check the type of the argument anyway.

Also, is `fetch()` on its own guaranteed to download the whole response? If `installFiles` calls `.arrayBuffer()` itself I'd like to be sure that nothing was waiting for that before reading the full data stream.

Not having to rely on an extension in the URL to indicate the format is important. The URL ending in `.tar` actually broke the site for a user because IDM was automatically hijacking the download, so I added an extra suffix. It also doesn't have a useful content type header at the moment. So unless you're going to inspect the contents of the raw bytes, I think being able to pass the format is important, regardless of how the rest of the API works.

I note you've written `.tar.gz` so I just want to reiterate that I'm using an uncompressed tarfile, no gzip. I'm letting compression happen at the HTTP level.

**Comment by hoodmane** [comment]
Thanks for the input @alexmojaki!

**Comment by grimmer0125** [comment]
> I think the possible advantage of passing ArrayBuffer instead of Promise is that exceptions raised by `fetch` can be easily handled outside of `pyodide.installFiles`. But I'm not a very JS person, so not sure it's worth it, maybe @grimmer0125 can share his idea?

If the total size of Python files is not large (e.g. <1MB), sequentially `await loadPyodide` and `await pyodide.installFile` (including download) may be acceptable since most of the time is counted for `Pyodide`. Parallel downloading Pyodide and user-defined Python files would be helpful for large Python file size or user preference.  

To achieve parallel downloading, `await pyodide.installFiles(tarPromise, installDirectory);` may be OK. But if this is the only way and passing tarPromise is the only possible parameter, it may not be much flexible. 

a) not everyone prefers `fetch`, for example, some people use [axios](https://github.com/axios/axios) to download data, and it has a different API to get the data.

b) a little tricky and it is not often seen that an API receives a `promise` parameter. 

`pyodide.installFilesFromArrayBuffer` is indeed flexible (pros.) and its disadvantage is similar to the promise way, it needs multiple steps. 

Besides these possibilities, I'm thinking of the other two things about achieving parallel downloading, 
1. End-user can consider using `Promise.all(fetch_job, loadPyodide );` which is similar to Python `asyncio.gather` to run concurrently and pass the fetch_result (e.g. `ArrayBuffer`) to `pyodide.installFiles`.
2. For those people who prefer to use a simpler API, an API like `pyodide.loadPyodideAndInstallFiles or pyodide.loadPyodide(extraInstallFiles)` may be good.

**Comment by hoodmane** [comment]
Okay so now I think we should have three options:
1. A way to install code without any dependency management or verification. This could be `installArchive` or something.
2. A way to install code with static dependency management (i.e., the dependency tree is pre-frozen outside of the browser). This is `loadPackage`. We could add a dependency overload system, but it would be preferable to give people good tools to fix the tree outside of the browser, and only fall back to dynamic overloading in weird cases.
3. A way to install code with dynamic dependency management (done inside the browser). This is micropip.

Maybe we could name this stuff better though...

**Comment by b4stien** [comment]
Hi everyone, at my company we're starting to use `pyodide` to share some code between backend and frontend and this prospect is amazing! Many thanks for the work.

For this matter we're a typical webapp with a Flask backend and a Typescript/React SPA frontend.

At that point we're struggling to understand the current best practice to ship backend code and execute it on the frontend, what we've seen discussed between here, the docs and some PRs :

- Build a custom Python distribution including our code from source.
  - Pros: it avoids any subsequent load over the network and the build can be pretty minimal.
  - Cons: it's super time consuming and is not feasible in a dev workflow.
- Build a Python wheel with our code on the backend, serve it to the front that loads `micropip` and `micropip` install this wheel.
  - Pros: it's documented and you can have a fast feedback loop.
  - Cons: you have to load micropip and we're not to sure about what's the best way to interact with our code afterwards, so far we're trying with some `pyodide.runPythonAsync()`.
- Use some APIs that we understand are the future of Pyodide but not really documented for now:
  - Directy ship our code to the front à la https://github.com/pyodide/pyodide/pull/1935#issuecomment-962835595 ?
  - Use a mixture of (is JS) `pyodide.unpackArchive()` and `pyodide.pyimport()` ?

With some guidance of what the community considers the best option I'd be glad to set up an "evergreen" example repository or propose some changes to the doc to reflect it.

**Comment by hoodmane** [comment]
> With some guidance of what the community considers the best option I'd be glad to set up an "evergreen" example repository or propose some changes to the doc to reflect it.

That would be great. The documentation is getting a bit out of date.

I think best practices are now something like:
```js
async function startup(){
	const pyodide_promise = loadPyodide(config);
	async function fetch_and_unpack(url) {
	      const fetch_promise = fetch(url).then(resp => resp.arrayBuffer());
	      const buffer = await fetch_promise;
	      const pyodide = await pyodide_promise;
	      pyodide.unpackArchive(buffer, url.substring(url.lastIndexOf('.') + 1));
	}
	await Promise.all(pkg_zip_wheel_or_tar_files_load.map(fetch_and_unpack));
    const pyodide = await pyodide_promise;
	// Do stuff.

}
```

**Comment by hoodmane** [comment]
Closing as completed.
